Meeting notes 2024-10-02.md (#687)
solid/specification: meetings/2024-10-02.md

### Participation and Code of Conduct
* [Join the W3C Solid Community Group](https://www.w3.org/community/solid/join), [W3C Account Request](http://www.w3.org/accounts/request), [W3C Community Contributor License Agreement](https://www.w3.org/community/about/agreements/cla/)
* [Solid Code of Conduct](https://github.com/solid/process/blob/main/code-of-conduct.md), [Positive Work Environment at W3C: Code of Conduct](https://www.w3.org/policies/code-of-conduct/)
* Operating principle for effective participation is to allow access across disabilities, across country borders, and across time. Feedback on tooling and meeting timing is welcome.
* If this is your first time, welcome! please introduce yourself.

### Scribes
* Michiel de Jong
* Sarven Capadisli

### Introductions
*

---

## Topics

### Solid Interop in Practice

URL: https://forum.solidproject.org/t/solid-interop-in-practice/7701

* eP: we mentioned this topic two weeks ago, this is a write-up by Rosano, who interviewed multiple Solid app developers. My main take-away is people find especially client-client interop hard to work with

> * eP: General comments on JS, there is a lot of work there but it's not specific to Solid.

Reviewer: I am not a developer but have been following Solid on Github for years. This conversation was interesting to me, so I engaged with ChatGPT on it to learn more about shapes and what the problem might be.  I thought the answers were interesting from the perspective of an outsider. I fed the entire conversation on this thread to chat. 

~~~~~~~~~~
You said:
Do you know what they mean by shapes?
~~~~~~~~~~~~
ChatGPT said:
ChatGPT
In the context of Solid (Social Linked Data) and RDF (Resource Description Framework), "shapes" refer to a way of defining the structure and constraints of RDF data. Specifically, shapes are part of a technology known as SHACL (Shapes Constraint Language), which is used to validate RDF data.

A shape defines expectations for the structure of data, such as what types of properties an RDF resource should have, the types of those properties, and other constraints like cardinality (e.g., how many times a property can occur). For instance, a "Person" shape might specify that a person has a name, an age, and an email, and that the name must be a string, the age an integer, and the email a valid email address.

Shapes are useful for ensuring interoperability between different applications that rely on RDF data. They provide a way to check if data conforms to a particular structure, making it easier for different systems or apps to work together by agreeing on a common "shape" of the data. In Solid, this could be used to make sure that data in a pod (personal online datastore) conforms to the expected structure that an application needs to function properly.

For example, if two apps share the same shape for an event or contact data, they can more easily read, write, and interpret each other's data, facilitating "interoperability" between Solid apps.

Does that clarify it?
~~~~~~~~~~
You said:
What seems to be the problem in this conversation re: shapes?
~~~~~~
ChatGPT said:
ChatGPT
The conversation around "shapes" in the Solid meetings seems to focus on a few key challenges:

1. Lack of Standardization and Collaboration on Shapes
There is a recurring theme in the discussion where participants note that app developers often have to create their own shapes, which adds a burden. The expectation seems to be that developers should not be left to define shapes independently. Instead, developers working in the same domain should collaborate to create a shared set of shapes (a version 1, for instance). This would reduce fragmentation and ensure greater interoperability. However, this kind of collaboration is not happening effectively, and there seems to be no easy or structured way for developers to agree on common shapes.

2. Domain-Specific vs. Generic Shapes
The conversation also points to an issue with defining what should be domain-specific shapes versus more generic ones. For example, there are certain kinds of data like calendar events or architectural blueprints that may require specific shapes unique to that domain. Participants argue about whether it’s best to create highly specialized shapes or to focus on more generic shapes that could work across different contexts. This balance between domain-specific needs and more general-purpose shapes seems unresolved.

3. Shape Discovery and Reuse
Another issue is the challenge of finding and reusing existing shapes. Some participants mention existing shape repositories, like the one on GitHub, but it's unclear how well these are being used or integrated. The lack of a clear process for discovering, reusing, or even sharing shapes seems to hinder interoperability. This suggests that there’s a lot of duplication of effort and inefficiency because developers might not know what shapes already exist or how to integrate them into their projects.

4. Composability and Nesting of Shapes
The conversation also touches on the technical complexity of making shapes more composable and nested. Participants mention the need to support more advanced use cases, like nesting shapes (for example, a shape within a shape). This could lead to complex requirements around access control and how different datasets interact with each other, which isn't being adequately addressed by current tools or protocols.

5. Process of Iterating on Shapes
Lastly, there seems to be an underlying issue with the process itself—how shapes should be iterated on, agreed upon, and updated. Some participants believe that there needs to be a better framework for "process and iterating," which would help standardize the work around shape definitions and make the development of shapes more systematic. The absence of this process means that current efforts are fragmented, leading to a slower path toward interoperability.

In summary, the conversation highlights the difficulty in establishing common shapes, the challenge of balancing generality with specificity, the lack of effective discovery and reuse mechanisms, and the need for a clearer process for shaping collaboration in Solid. The current approach is ad-hoc, which leads to fragmented solutions that undermine the goal of interoperability between Solid applications.
~~~~~~~
You said:
What would you recommend the approach be?
~~~~~~~~

ChatGPT said:
ChatGPT
To address the challenges discussed in the Solid conversation around "shapes," a structured approach could help foster better collaboration, standardization, and interoperability. Here are some key recommendations:

1. Establish a Centralized Shape Registry
A centralized and well-maintained Shape Repository (or registry) should be created, where all shared shapes are published, documented, and easily accessible to developers. This would include:

Versioning: Allowing different versions of shapes to evolve over time while maintaining backward compatibility.
Metadata: Clearly labeling the domain, purpose, and constraints of each shape (e.g., for calendar events, architectural blueprints, etc.).
Discoverability: Ensuring developers can easily search for existing shapes based on domain, use case, or data type, avoiding duplication of effort.
The key is to make it easy for developers to reuse existing shapes and contribute back, reducing fragmentation.

2. Define a Process for Shape Standardization
Introduce a shape standardization process similar to how W3C or other technical bodies define standards:

Community Collaboration: Establish working groups (or task forces) around specific domains (e.g., social networking, event planning, healthcare) where developers working on similar applications can come together to define shapes.
Consensus-Based Approach: Allow developers to propose shapes, but ensure a community review process before finalizing them. This prevents isolated or redundant shape definitions.
Interoperability Testing: Before shapes are added to the central registry, run tests to ensure they work in different applications or contexts to validate interoperability.
This would formalize how shapes are defined, refined, and adopted, reducing the burden on individual developers.

3. Encourage Reusability Through Modular Shapes
To balance between generic and domain-specific needs, promote modular, composable shapes. Shapes should be designed in a way that:

Core/Generic Shapes: Define core shapes that represent general concepts (e.g., person, event) with a minimal set of constraints.
Extensions/Domain-Specific Shapes: Allow domain-specific shapes (e.g., medical appointment) to build on top of these core shapes by adding specific constraints or properties. This will enable reuse across domains while still supporting domain-specific requirements.
By creating building blocks of shapes, developers can customize existing ones for their specific needs without starting from scratch.

4. Tooling for Shape Validation and Generation
Provide tools that make using and creating shapes easier:

Shape Validation Tools: Offer easy-to-use tools that allow developers to validate RDF data against a shape. This will give developers confidence that their data conforms to the expected structure and can be shared across applications.
Shape Generators: Develop tools that generate shapes from existing RDF data or ontology. For example, if a developer already has a data model in use, they could generate the corresponding shape and refine it collaboratively.
These tools would lower the technical barrier to adopting shapes and encourage broader participation from developers who are not RDF or SHACL experts.

5. Facilitate Domain-Specific Communities
Encourage the formation of domain-specific working groups within the Solid community. These groups can:

Agree on Shared Shapes: Ensure collaboration among developers working on similar applications to define and agree on common shapes for their domain (e.g., education, finance, healthcare).
Shape Governance: Create a lightweight governance model for how shapes are proposed, reviewed, and iterated on, similar to how standards bodies like W3C operate.
Shape Review Process: Establish a mechanism for feedback and improvement of existing shapes over time based on real-world use cases.
Facilitating these communities would allow domain experts to bring their knowledge into the process and align on best practices.

6. Develop Use Case-Driven Shape Proposals
Instead of proposing shapes in a vacuum, ensure they are use-case driven:

Real-World Applications: Shape proposals should come directly from the needs of real-world applications or communities that are using Solid. This ensures that shapes are relevant and directly tied to practical challenges developers face.
Pilot Projects: Run pilot projects with willing participants where new shapes are proposed and tested in real applications before they are made part of the standard set.
This approach would ensure that the shapes being developed solve real interoperability issues in practice, rather than being theoretical.

7. Shape Composition and Relations
Address the need for composability and nested shapes by encouraging:

Shape Inheritance: Allow shapes to inherit properties from other shapes. For example, a "MedicalAppointment" shape could inherit from a generic "Event" shape but add specific fields for doctor’s name, prescription, etc.
Context-Aware Shapes: Develop patterns for how shapes are applied depending on the context, especially for access control. This is critical for sensitive data (e.g., health records) where different data consumers might need different levels of access.
Building these patterns into Solid’s shape system would make them more flexible and powerful for real-world use.

8. Shape-First Approach for Interoperability
Finally, promote a shape-first approach to interoperability:

Shared Shape Libraries: Encourage developers to start by defining or choosing shapes before developing applications. If apps are built around shared shapes, interoperability becomes easier to achieve.
Shape Validation as Part of App Development: Make shape validation an integral part of app development, so developers ensure compliance with agreed-upon structures early in the development process.
This approach would shift the focus toward more proactive alignment on data structures, leading to greater interoperability across Solid apps.

By following these recommendations, the Solid community can move toward a more standardized, collaborative approach to defining and using shapes, addressing the current fragmentation and interoperability challenges.

> * eP: General comments on JS, there is a lot of work there but it's not specific to Solid.

Reviewer 2: @paulduplantis — 

I suggest you put this comment on a higher level, either as [a new issue](https://github.com/solid/specification/issues/) since it doesn't seem like your comments are tied to the meeting for which these are the minutes, or as a comment [on this PR](https://github.com/solid/specification/pull/687) instead of as a comment on a change suggestion on this PR. 

When you do this, I strongly suggest that you do not put such large chunks of the content into code blocks, as they are very hard to read, requiring manual side-scrolling for each VERY long line of text. I suggest you use the `>` character to indicate quoted sections, using two of them `>>` for your prompts and one of them `>` for ChatGPT's responses.

* eP: Coming back to LWS UCs re overlap. As I understand hte harter, the client-client is out of scope. There is not going to be domain-specific. Naturally the WG will not fully satisfy all the UCs because there is always the client-client part. That's why I think the UCs should be shared. Ensuring what we have fits into the LWS.
* RG: There can be generic use cases, I'll point you to Douglas Engelbart in that respect. Use cases don't necessarily have to be domain-specific. We have a generic UC on what knowledge-workshop would be and thats a good light of how software should be designed.
* MdJ: To me the topic of how different specs are interoperable is interesting. As an app dev, one might use one and dev another. If all those things are incubating all being interop under Solid.

### Use cases

URL: https://solid.github.io/authorization-panel/authorization-ucr/
URL: https://github.com/solid/user-stories
URL: https://github.com/solid/webid-profile/blob/main/notes/use-cases.md
URL: https://solid.github.io/notifications-panel/notifications-ucr

> * eP: Authorization use case were written in reverse, starting from WAC as a solution plus some know limitations.
> * eP: Instead of making up use cases, we can work with existing communities, preferably open source and document their real world usecases which already have user base. As an example we had a meeting with https://openfoodnetwork.org/
> * eP: We can also try to "eat our own cookings" more. This prototype tries to define some CG meetings based requirements https://github.com/janeirodigital/sai-js/tree/main/examples/plenary#requirements

Reviewer 2: ```suggestion
* eP: Authorization use case were written in reverse, starting from WAC as a solution, plus some known limitations.
* eP: Instead of making up use cases, we can work with existing communities, preferably open source, and document their real world use cases, which already have a user base. As an example, we had a meeting with <https://openfoodnetwork.org/>
* eP: We can also try to "eat our own cooking" more. This prototype tries to define some CG meetings based requirements: <https://github.com/janeirodigital/sai-js/tree/main/examples/plenary#requirements>.
```

* eP: Unfortunately I didn't participate in the TPAC UC meeting.
* eP: We (CG) has a bunch of work on this already (linked above).

> * eP: We need to have UCR managed in a way to evaluate proposals. Do we do it by simplicity or emotions or something else? ??? objectively. We can have a matrix for the proposals, then we can see best on practice. With e.g., added complexity, we could do with simpler. This is because different people coming with different proposals. Once we have implementation experience it is going to be subjective. We can have some rough way of evaluating the complexity. For me the main point for the UCs is to evaluate all the proposals. We are kind of stuck on what to focus on.
> * RG: I like a lot of the ideas behind on having some process to make a note of it, but at the end of the day we could do the matrices etc but it'll be subjective and no real way of measuring complexity. It is more of art than science.

Reviewer 2: ```suggestion
* eP: We need to have UCR managed in a way to evaluate proposals. Do we do it by simplicity, or emotions, or something else? ??? objectively. We can have a matrix for the proposals, then we can see best on practice. With, for instance, added complexity, we could do with simpler. This is because different people come with different proposals. Once we have implementation experience, it is going to be subjective. We can have some rough way of evaluating the complexity. For me, the main point for the UCs is to evaluate all the proposals. We are kind of stuck about what to focus on.
* RG: I like a lot of the ideas about having some process to make a note of it, but at the end of the day we could do the matrices, etc., but it'll be subjective and no real way of measuring complexity. It is more of an art than a science.
```

> * MdJ: We can measure complexity by looking at how many implementing. Adding traction as a criteria.
> * RG: That's one measure and so many different ways of measuring. It is not something we can solve for, but to have an intuition for.
> * MdJ: If you look to see what solutions get traction, that is a proxy for how complex they are to implement. Look at shape files for instance, very few app devs write them, I think that's an indication that it's quite a complex solution.

Reviewer 2: ```suggestion
* MdJ: We can measure complexity by looking at how many are implementing. Add traction as a criteria.
* RG: That's one measurement with many different ways of being measure. It is not something we can solve for, but can have an intuition for.
* MdJ: If you look to see what solutions get traction, that is a proxy for how complex they are to implement. Look at shape files, for instance; very few app devs write them. I think that's an indication that it's quite a complex solution.
```

> * SC: Use cases are important. In the past we had a use case survery, which ties to Michiel's point. This can signal from the community interest and commitment. People are brining up age old issues about Solid should be doing this or that. We have an aboundance of use cases. Data portability comes up, access requests. We had issues, we had proposals. For example [data migrator](https://pdsinterop.org/solid-migrator/). I'm also part of this and we didn't do great job on evaluating those proposals. We need storage provisiooning, migrating storate. There are proposals and we don't give them enough attention. Later in the future we come back "oh we need that". We do need those use cases but there are solutions in front of us. A solution is better than no solution. Recently we heard need for digital wallets, sometimes it is hard to dig out those proposals. We need to be evaluating those solutions, if we don't them enough attention people get impression that we don't take them seriously. I don't want us to get lost in the process and loos track of simple solutions in front of us.

Reviewer 2: ```suggestion
* SC: Use cases are important. In the past we had a use case survey, which ties to Michiel's point. This can signal interest and commitment from the community. People are bringing up age old issues about "Solid should be doing this or that". We have an abundance of use cases. Data portability comes up, access requests. We had issues, we had proposals. For example, [data migrator](https://pdsinterop.org/solid-migrator/). I'm also part of this, and we didn't do great job on evaluating those proposals. We need storage provisioning, storage migration. There are proposals, and we don't give them enough attention. Later in the future we come back, "oh, we need that". We do need those use cases, but there are solutions in front of us. A solution is better than no solution. Recently we heard a need for digital wallets; sometimes it is hard to dig out those proposals. We need to be evaluating those solutions. If we don't give them enough attention, people get the impression that we don't take them seriously. I don't want us to get lost in the process and lose track of simple solutions in front of us.
```

> * eP: I hear there is an agreement on evaluating proposal. Also agree with SC on signal from community and what to implement. If UCs are real-world, then if a UC is not mine but an alternative is like mine, then I can use those three requirements. We can look at more generic and specific level. (from chat: example - https://github.com/solid/authorization-panel/tree/main/proposals/evaluation)
> * eP: re MdJ, we need to have a little more signaling. How the dev chooses one of the alternative proposals. With a matrix the dev can say I also have those requirements so I'll try to implement or 80% is good enough. Other people can make educated guesses on what to use. We also need the matrix of which proposal would allow implementers to decide. Also to know what they like or don't like.

Reviewer 2: ```suggestion
* eP: I hear an agreement on evaluating proposals. Also agree with SC on signals from community about what to implement. If UCs are real-world, then if a UC is not mine but an alternative is like mine, then I can use those three requirements. We can look at more generic and specific level. (from chat: example — <https://github.com/solid/authorization-panel/tree/main/proposals/evaluation>)
* eP: re MdJ, we need to have a little more signaling. How the dev chooses one of the alternative proposals. With a matrix, the dev can say, "I also have those requirements so I'll try to implement" or "80% is good enough". Other people can make educated guesses on what to use. We also need the matrix of proposals to allow implementers to decide. Also to know what they like or don't like.
```

* MdJ: Also need people to vote with their feet.

> * RG: SC, one of the things is that all of these discusssions are very buried. I try to stab at one or two of these things, but we really need some effort to bring all of those from GH. You have this incredible domain knowledge due to you having been here for years but most of us are newer. There has to be more visibility on those discussions.
> * PAC: To react to SC, taking into account previous work, i think it is very important. It might be very importnt in the LWS WG for each UC. But also maybe reference earlier work in order to see what the CG has done with the UCs. So, ack not at all starting from scratch.

Reviewer 2: ```suggestion
* RG: SC, one of the things is that all of these discusssions are very buried. I try to stab at one or two of these things, but we really need some effort to bring all of those from GH. You have this incredible domain knowledge, due to you having been here for years, but most of us are newer. There has to be more visibility on those discussions.
* PAC: To react to SC, I think it is very important to take previous work into account. It might be very important in the LWS WG for each UC. But also, maybe reference earlier work, to see what the CG has done with the UCs. So, not at all starting from scratch.
```

> * eP: Agree with PAC on crowdsourcing. I think it is fair to put the burden on the proposer to mark what requirements it is addressing.
> * PAC: I agree that we have a living UCs for the CG. For the WG, there is more of a deadline driven UCs. Also to decide how we design the REC. Both are valuable. Please don't expect the WG to drive the UCs. The scope of the WG is narrower. There should be between the two work. LWS would have a subset of UCs.
> * eP (in chat): the dataset can aggregate use cases and requirements from the WG. the work Jeff does on the catalog and me on solid efforts can provide some foundations for it 

Reviewer 2: ```suggestion
* eP: Agree with PAC on crowdsourcing. I think it is fair to put the burden on the proposer to mark what requirements it is addressing.
* PAC: I agree that we have a living UC for the CG. For the WG, there is more of a deadline-driven UC. Also to decide how we design the REC. Both are valuable. Please don't expect the WG to drive the UCs. The scope of the WG is narrower. There should be work between the two. LWS would have a subset of UCs.
* eP (in chat): the dataset can aggregate use cases and requirements from the WG. the work Jeff does on the catalog and me on solid efforts can provide some foundations for it 
```
